std/crypto/tls/conn.go

```diff
@@ -1574,10 +1591,14 @@ func (c *Conn) release() {
 
 // Close closes the connection.
 func (c *Conn) Close() error {
-	c.closeMux.Lock()
+	if !c.hasLock {
+		c.closeMux.Lock()
+	}
 	closed := c.closed
 	c.closed = true
-	c.closeMux.Unlock()
+	if !c.hasLock {
```

Comment:
2. another goroutine set c.hasLock = false before we get c.hasLock again.
3. we get a `c.hasLock == false` and call c.closeMux.Unlock(), without the previos Lock(), then we went wrong.